auto_round/autoround.py

```diff
@@ -506,48 +506,84 @@ def _set_device_for_matching_module(self, name: str, device: str) -> None:
+        block_memory, input_ouput_memory = self.get_block_info(
+            block, input_ids
         )
+        
+        mem_expansion_factor = 13 if self.mem_expansion_factor is None else self.mem_expansion_factor
 
-        cuda_devices = [f"cuda:{i+1}" for i in range(num_gpus)]
+        if block_memory * mem_expansion_factor < device_0_memory:
+            return  # fit in one GPU
 
-        cnt = 0
+        cuda_devices = [f"cuda:{i}" for i in range(num_gpus)] # only support cuda now
         device_map = {}
-        # Iterate through all expert Linear layers and assign them to GPUs
-        for n, m in self.model.named_modules():
-            if hasattr(type(m), "__name__") and "ModuleList" in type(m).__name__:
-                cnt += 1  # find the first modulelist
-                if cnt == 2:  # find the second modulelist
-                    cnt -= 1
-                    for i in range(device_0_block_num, block_num):
-                        for name, module in m[i].named_modules():
-                            if isinstance(module, torch.nn.Linear):
-                                full_gpu_block = (block_num // num_gpus) * num_gpus
-                                if i >= full_gpu_block:
-                                    # If the number of blocks exceeds the number of GPUs,
-                                    # distribute the remaining blocks in multiples of the number of GPUs
-                                    device_index = i % num_gpus
-                                else:
-                                    device_index = i // (block_num // num_gpus) % num_gpus
-                                device_map.update({n + "." + str(i) + "." + name: cuda_devices[device_index]})
-        self.device_map = device_map
-        self._set_device_map_in_blocks(self.device_map)
+        device_memory = {device: get_device_memory(index) for index, device in enumerate(cuda_devices)}
+        device_memory["cuda:0"] = device_0_memory - input_ouput_memory
+
+        device_idx = 0
+        for n, m in block.named_modules():
+            if isinstance(m, supported_types):
+                layer_name = block.tmp_name + "." + n
+                layer_memory = sum(p.numel() for p in m.parameters()) * m.weight.element_size() / 1024**3
+                if layer_memory * mem_expansion_factor < device_memory[cuda_devices[device_idx]]:
+                    device_map[layer_name] = cuda_devices[device_idx]
+                    device_memory[cuda_devices[device_idx]] -= layer_memory * mem_expansion_factor
+                else:
+                    device_idx += 1
+                    device_map[layer_name] = cuda_devices[device_idx]
+                    device_memory[cuda_devices[device_idx]] -= layer_memory * mem_expansion_factor
+                    if device_idx >= len(cuda_devices):
+                        raise ValueError(
+                            f"model is too large to fit in {num_gpus} GPUs, "
```

Comment:
as discussed, for device 0, we use the mem_expansion_factor, for other devices, we just split the remaining parameters. If it's more than the  layer_memory * mem_expansion_factor, logger a warning but not an exception